segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerRecoveryUtils.java

```diff
@@ -0,0 +1,210 @@
+     * All segments from the storage are listed one by one, then mapped to their corresponding {@link DebugStreamSegmentContainer}
+     * instances for registering them to container metadata segment.
+     * @param storage                           A {@link Storage} instance that will be used to list segments from.
+     * @param debugStreamSegmentContainers      A Map of Container Ids to {@link DebugStreamSegmentContainer} instances
+     *                                          representing the containers that will be recovered.
+     * @param executorService                   A thread pool for execution.
+     * @throws InterruptedException             Required for Futures.get()
+     * @throws ExecutionException               Required for Futures.get()
+     * @throws TimeoutException                 Required for Futures.get()
+     * @throws IOException                      Requited for Storage.listSegments()
+     */
+    public static void recoverAllSegments(Storage storage, Map<Integer, DebugStreamSegmentContainer> debugStreamSegmentContainers,
+                                          ExecutorService executorService) throws InterruptedException, ExecutionException,
+            TimeoutException, IOException {
+        Preconditions.checkNotNull(storage);
+        Preconditions.checkNotNull(executorService);
+        Preconditions.checkNotNull(debugStreamSegmentContainers);
+        Preconditions.checkArgument(debugStreamSegmentContainers.size() > 0, "There should be at least one " +
+                "debug segment container instance.");
+
+        log.info("Recovery started for all containers...");
+        // Get all segments in the container metadata for each debug segment container instance.
+        Map<Integer, Set<String>> existingSegmentsMap = getExistingSegments(debugStreamSegmentContainers, executorService);
+
+        SegmentToContainerMapper segToConMapper = new SegmentToContainerMapper(debugStreamSegmentContainers.size());
+
+        Iterator<SegmentProperties> segmentIterator = storage.listSegments();
+        Preconditions.checkNotNull(segmentIterator);
+
+        // Iterate through all segments. Create each one of their using their respective debugSegmentContainer instance.
+        ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
+        while (segmentIterator.hasNext()) {
+            val currentSegment = segmentIterator.next();
+
+            // skip recovery if the segment is an attribute segment.
+            if (NameUtils.isAttributeSegment(currentSegment.getName())) {
+                continue;
+            }
+
+            int containerId = segToConMapper.getContainerId(currentSegment.getName());
+            existingSegmentsMap.get(containerId).remove(currentSegment.getName());
+            futures.add(recoverSegment(debugStreamSegmentContainers.get(containerId), currentSegment));
+        }
+        Futures.allOf(futures).join();
```

Comment:
Here you use `join` but in other methods you use `get(Timeout)`. Please be consistent. If you already use `get(Timeout)` in one place, you might as well use it everywhere.

Reply:
Using get now.